ares/n64/cpu/memory.cpp

```diff
@@ -168,132 +169,40 @@ inline auto CPU::busReadBurst(u32 address, u32 *data) -> void {
   return bus.readBurst<Size>(address, data, *this);
 }
 
-auto CPU::fetch(u64 vaddr) -> maybe<u32> {
-  if(vaddrAlignedError<Word>(vaddr, false)) return nothing;
-  switch(segment(vaddr)) {
-  case Context::Segment::Unused:
-    step(1 * 2);
-    addressException(vaddr);
-    exception.addressLoad();
-    return nothing;
-  case Context::Segment::Mapped:
-    if(auto match = tlb.load(vaddr)) {
-      if(match.cache) return icache.fetch(vaddr, match.paddr & context.physMask, cpu);
-      step(1 * 2);
-      return busRead<Word>(match.paddr & context.physMask);
-    }
-    step(1 * 2);
-    addressException(vaddr);
-    return nothing;
-  case Context::Segment::Cached:
-    return icache.fetch(vaddr, vaddr & 0x1fff'ffff, cpu);
-  case Context::Segment::Cached32:
-    return icache.fetch(vaddr, vaddr & 0xffff'ffff, cpu);
-  case Context::Segment::Direct:
-    step(1 * 2);
-    return busRead<Word>(vaddr & 0x1fff'ffff);
-  case Context::Segment::Direct32:
-    step(1 * 2);
-    return busRead<Word>(vaddr & 0xffff'ffff);
-  }
-
-  unreachable;
+auto CPU::fetch(PhysAccess access) -> maybe<u32> {
+  step(1 * 2);
+  if(!access) return nothing;
+  if(vaddrAlignedError<Word>(access.vaddr, false)) return nothing;
+  if(access.cache) return icache.fetch(access.vaddr, access.paddr, cpu);
+  return busRead<Word>(access.paddr);
 }
 
 template<u32 Size>
-auto CPU::read(u64 vaddr) -> maybe<u64> {
-  if(vaddrAlignedError<Size>(vaddr, false)) return nothing;
-  GDB::server.reportMemRead(vaddr, Size);
-  
-  switch(segment(vaddr)) {
-  case Context::Segment::Unused:
-    step(1 * 2);
-    addressException(vaddr);
-    exception.addressLoad();
-    return nothing;
-  case Context::Segment::Mapped:
-    if(auto match = tlb.load(vaddr)) {
-      if(match.cache) return dcache.read<Size>(vaddr, match.paddr & context.physMask);
-      step(1 * 2);
-      return busRead<Size>(match.paddr & context.physMask);
-    }
-    step(1 * 2);
-    addressException(vaddr);
-    return nothing;
-  case Context::Segment::Cached:
-    return dcache.read<Size>(vaddr, vaddr & 0x1fff'ffff);
-  case Context::Segment::Cached32:
-    return dcache.read<Size>(vaddr, vaddr & 0xffff'ffff);
-  case Context::Segment::Direct:
-    step(1 * 2);
-    return busRead<Size>(vaddr & 0x1fff'ffff);
-  case Context::Segment::Direct32:
-    step(1 * 2);
-    return busRead<Size>(vaddr & 0xffff'ffff);
-  }
-
-  unreachable;
+auto CPU::read(PhysAccess access) -> maybe<u64> {
```

Comment:
It looks like we lost the step() calls entirely here. Same applies to CPU::write.

Reply:
All right. We should call out that this change affects timing, and it will be no surprise if we see an impact in the usual suspects (e.g. DK64).